apps/aevm/src/aevm_eeevm.erl

```diff
@@ -1424,3 +1408,92 @@ log_topics(AccountAddress, Topics) ->
+    State0            = spend_call_gas(StateIn, Op),
+
+    {Gas, State1}     = pop(State0),
+    {To, State2}      = pop(State1),
+    {Value, State3}   = case Op of
+                            ?CALL         -> pop(State2);
+                            ?CALLCODE     -> pop(State2);
+                            ?DELEGATECALL -> {aevm_eeevm_state:value(State2), State2}
+                        end,
+    {IOffset, State4} = pop(State3),
+    {ISize, State5}   = pop(State4),
+    {OOffset, State6} = pop(State5),
+    {OSize, State7}   = pop(State6),
+    Dest              = case Op of
+                            ?CALL -> To;
+                            ?CALLCODE -> aevm_eeevm_state:address(State6);
+                            ?DELEGATECALL -> aevm_eeevm_state:address(State6)
+                        end,
+    {I, State8}       = aevm_eeevm_memory:get_area(IOffset, ISize, State7),
+    GasAfterSpend     = aevm_eeevm_state:gas(State8),
+    Code              = aevm_eeevm_state:extcode(To, State8),
+    CallGas = case Value =/= 0 of
+                  true -> ?GCALLSTIPEND + Gas;
+                  false -> Gas
+              end,
+    case GasAfterSpend >= CallGas of
+        true  -> ok;
+        false -> eval_error(out_of_gas, State7)
+    end,
+    Caller = case Op of
+                 ?CALL -> aevm_eeevm_state:address(State8);
+                 ?CALLCODE -> aevm_eeevm_state:address(State8);
+                 ?DELEGATECALL -> aevm_eeevm_state:caller(State8)
+             end,
+    CallState = aevm_eeevm_state:prepare_for_call(Caller, Dest, CallGas, Value,
+                                                  Code, CallDepth,
+                                                  State8),
+    case aevm_eeevm_state:no_recursion(State8) of
+        true  -> %% TODO: Set callcreates
+            GasOut = GasAfterSpend + CallGas,
+            State9 = aevm_eeevm_state:set_gas(GasOut, State8),
+            State10 = aevm_eeevm_state:add_callcreates(#{ data => I
+                                                        , destination => Dest
+                                                        , gasLimit => CallGas
+                                                        , value => Value
+                                                        }, State9),
+            {1, State10};
+        false ->
+            %% TODO: Should this not count towards mem gas?
+            CallState1 = aevm_eeevm_memory:write_area(0, I, CallState),
+            {OutGas, OutState, R} =
+                case eval(CallState1) of
+                    {ok, OutState2} -> {aevm_eeevm_state:gas(OutState2), OutState2, 1};
+                    {error, out_of_gas,_RState2} -> {0, CallState1, 1}
+                end,
+            CallTrace = aevm_eeevm_state:trace(OutState),
+            %% Go back to the caller state.
+            ReturnState1 = aevm_eeevm_state:add_trace(CallTrace, State8),
+            GasAfterCall = GasAfterSpend + OutGas,
```

Comment:
`GasAfterSpend + OutGas - CallGas`?

Reply:
No. GasAfterSpend is how much gas you would have if all the gas you give to the caller is used up. So we add back the unused amount after the call.

Reply:
`GasAfterSpend` is computed before `CallGas`, so I don't see how that could be. Also, if that is really the case then the check `GasAfterSpend >= CallGas` makes no sense at all.

As far as I can tell `GasAfterSpend` is the amount of gas you have after paying for the call instruction, computed by `spend_call_gas`.

Reply:
It is not so easy to see...
spend_call_gas includes the dynamic cost of the call, that is also the amount of gas you transfer to the call (i.e. the gas argument to the instruction). See aevm_gas.erl line 80 and forward.

Reply:
The check look suspicious though.

Reply:
Rewritten to check for 0 and to not give the call stipend back.
Things should be closer to right now but this is just the beginning.
When we can hook in the chain for real we can start doing some real tests, and fix things like handling calls to new contracts.